onnxruntime/core/framework/session_state.cc

```diff
@@ -1526,21 +1543,33 @@ static void BindToDeviceStream(const SequentialExecutionPlan& execution_plan,
 }
 
 std::unique_ptr<DeviceStreamCollection> SessionState::AcquireDeviceStreamCollection() const {
-  std::lock_guard<onnxruntime::OrtMutex> lock(device_stream_pool_mutex_);
-  if (!device_stream_pool_.empty()) {
-    auto device_stream = std::move(device_stream_pool_.back());
-    device_stream_pool_.pop_back();
-    return device_stream;
+  if (has_device_stream_enabled_ep_) {
+    std::lock_guard<onnxruntime::OrtMutex> lock(device_stream_pool_mutex_);
+    if (!device_stream_pool_.empty()) {
+      auto device_stream = std::move(device_stream_pool_.back());
+      device_stream_pool_.pop_back();
+      return device_stream;
+    } else {
+      auto device_stream = std::make_unique<DeviceStreamCollection>(this->GetExecutionPlan()->execution_plan.size(), *this);
+      BindToDeviceStream(*this->GetExecutionPlan(), *device_stream, *stream_handles_registry_);
+      return device_stream;
+    }
   } else {
+    // no reusing of device stream is needed, just return a new instance.
     auto device_stream = std::make_unique<DeviceStreamCollection>(this->GetExecutionPlan()->execution_plan.size(), *this);
```

Comment:
> std::make_unique<DeviceStreamCollection> [](http://example.com/codeflow?start=24&length=41)

Is there anyway we can remove allocations?

Reply:
update the implementation to avoid instantiate the DeviceStreamCollection when there is no EP use stream